File: frontend/src/types/websocket.ts
import type { ScriptResult } from '@/types/execute'
import type { WsFormat } from '@/lib/ws-settings'

export type WsDir = 'in' | 'out' | 'system'

// Store-side status (what the editor renders).
export type WsStatus = 'disconnected' | 'connecting' | 'connected' | 'closing' | 'error'

// Backend connection states emitted on `ws:state:<id>` — these mirror
// ConnState.String() in Go ('closed', not 'disconnected').
export type WsBackendState = 'connecting' | 'connected' | 'closing' | 'closed' | 'error'

export interface WsMessage {
  id: string
  dir: WsDir
  ts: number
  data: string
  format?: WsFormat
  level?: 'error' // system row rendered in red
  failed?: boolean // outgoing row marked failed when Send errors
}

// Connect RPC result: a refused handshake arrives here, not as an error, so the
// pre-connect script output survives with it.
export interface WsConnectResult {
  connected: boolean
  connectionId: string
  status: number
  subprotocol: string
  error?: string
  script?: ScriptResult
}

// Payload of the `ws:message:<id>` Wails event.
export interface WsIncoming {
  dir: WsDir
  data: string
  type: 'text' | 'binary'
  at: number
}

// Payload of the `ws:state:<id>` Wails event.
export interface WsStateEvent {
  state: WsBackendState
  error?: string
}

File: internal/adapters/wails/dto/websocket.go
package dto

// The connection id comes from the frontend so it can subscribe before the handshake.
type WSConnectRequest struct {
	RequestID    string `json:"requestId"`
	WorkspaceID  string `json:"workspaceId"`
	ConnectionID string `json:"connectionId"`
}

// A failed handshake is reported here, not an error, so pre-connect output survives.
type WSConnectResultDTO struct {
	Connected    bool             `json:"connected"`
	ConnectionID string           `json:"connectionId"`
	Status       int              `json:"status"`
	Subprotocol  string           `json:"subprotocol"`
	Error        string           `json:"error,omitempty"`
	Script       *ScriptResultDTO `json:"script,omitempty"`
}

// Binary payloads travel base64-encoded.
type WSSendRequest struct {
	ConnectionID string `json:"connectionId"`
	Data         string `json:"data"`
	MessageType  string `json:"messageType"` // "text" | "binary"
}

type WSDisconnectRequest struct {
	ConnectionID string `json:"connectionId"`
}

// WSMessageDTO is emitted on `ws:message:<connID>`.
type WSMessageDTO struct {
	Dir  string `json:"dir"` // "in" | "out" | "system"
	Data string `json:"data"`
	Type string `json:"type"` // "text" | "binary"
	At   int64  `json:"at"`   // unix milliseconds
}

// WSStateDTO is emitted on `ws:state:<connID>`.
type WSStateDTO struct {
	State string `json:"state"` // connecting|connected|closing|closed|error
	Error string `json:"error,omitempty"`
}
